Automattic/newspack-ads
# Add GAM integration

### Changes proposed in this Pull Request:



The ad units created in GAM have to be manually recreated in Newspack, linking one with the other via the unique ad placement code. With changes here, the ad units are fetched from GAM and saved in the DB - in order not to query GAM on each page load.

The only downside is that the sync happens when the user edits the ad units, so the changes made in [GAM web UI](https://admanager.google.com/) will not be reflected until the user logs in to the Advertising wizard of Newspack plugin. This is not a regression though - it was like that before. 

### How to test the changes in this Pull Request:

1. On `master`, set up some ad units to show (via global placements or the Ad block) - to ensure the update will not break existing ad unit placements
1. Switch to this branch, and use `add/gam-ad-units-in-ads-plugin` branch of the Newspack plugin (https://github.com/Automattic/newspack-plugin/pull/936)
2. Do not authenticate via Google yet (or [revoke access](https://myaccount.google.com/permissions))
3. Visit the site, verify that the ads work as before
4. Visit the Advertising wizard - observe that legacy ad units are listed and editable
5. Set `NEWSPACK_ADS_USE_GAM` env var to value `true` - this enables the GAM connection
1. Authenticate Newspack to use Google data via Newspack dashboard (`NEWSPACK_GOOGLE_OAUTH_ENABLED` env var has to be set)
2. Observe that now the ad units like displays both legacy ad units, and GAM ad units
3. Create, edit, then archive a GAM ad unit - observe the changes are reflected in [GAM web UI](https://admanager.google.com/) at each step
4. Assign an ad unit to a global placement or insert via the Ad block - observe the ad unit displaying an ad* 
5. Observe the ad units set up while on `master` are still displayed
6. Change the network code in ~the options table (`_newspack_ads_service_google_ad_manager_network_code`)~ _edit: change value of `$user_network_code` [in `is_network_code_matched` method](https://github.com/Automattic/newspack-ads/blob/c30a9dd307f4aba23ad4e42d4591943e7189b8d3/includes/class-newspack-ads-model.php#L350)_ to something else. Observe a notice about the mismatch above ad units, and that the ad units editing and creation is unavailable in the UI

\* if properly configured - but in non-AMP context, an ad "surface" will be displayed anyway
